### PR TITLE
auto enable webhook

### PR DESCRIPTION
**Pull request by jampulanaveen**

### CODE REVIEW

File: woo-razorpay.php [hunk 1]
```diff
@@ -315,9 +287,22 @@ public function autoEnableWebhook()
...
             $key_id      = $this->getSetting('key_id');
             $key_secret  = $this->getSetting('key_secret');
-            $enabled     = $this->getSetting('enable_webhook');
-            $secret      = $this->getSetting('webhook_secret');
+            $enabled     = 'yes';
+            $secret      =  bin2hex(openssl_random_pseudo_bytes(4));
+
+            $getWebhookFlag =  get_option('webhook_enable_flag');
+            
+            $TimeNow = new DateTime('now');
```
Comment (ChetanGN) [inline woo-razorpay.php:295]: @jampulanaveen Make it camelcase.

File: woo-razorpay.php [hunk 2]
```diff
@@ -315,9 +287,22 @@ public function autoEnableWebhook()
...
+            $time =  $TimeNow->getTimestamp();
 
+            if(empty($getWebhookFlag)){
```
Comment (ChetanGN) [inline woo-razorpay.php:298]: @jampulanaveen Maintain code format.

File: woo-razorpay.php [hunk 3]
```diff
@@ -610,6 +573,26 @@ private function getRedirectUrl($orderId)
...
+                $current  = new DateTime('now');
+                $currentTime = $current->format('Y-m-d H:i:s');
+                
+                $getCurrentTime = new DateTime($currentTime);
+            
+                $interval = $webhookTime->diff($getCurrentTime); 
+            
+                $hours = ($interval->days * 24) + $interval->h;
+                
+
```
Comment (ChetanGN) [inline woo-razorpay.php:590]: Remove blank space

File: woo-razorpay.php [hunk 4]
```diff
@@ -610,6 +573,26 @@ private function getRedirectUrl($orderId)
...
+                $webhookTime = new DateTime(date('Y-m-d H:i:s', $getWebhookFlag));
+                $current  = new DateTime('now');
+                $currentTime = $current->format('Y-m-d H:i:s');
+                
+                $getCurrentTime = new DateTime($currentTime);
+            
+                $interval = $webhookTime->diff($getCurrentTime); 
+            
+                $hours = ($interval->days * 24) + $interval->h;
```
Comment (ramth05) [inline woo-razorpay.php:600]: @jampulanaveen, We don't needs hours conversion here, just use time stamp + 24 hours ( in seconds) for comparision.

File: woo-razorpay.php [hunk 5]
```diff
@@ -315,9 +287,22 @@ public function autoEnableWebhook()
 
             $key_id      = $this->getSetting('key_id');
             $key_secret  = $this->getSetting('key_secret');
-            $enabled     = $this->getSetting('enable_webhook');
-            $secret      = $this->getSetting('webhook_secret');
+            $enabled     = 'yes';
+            $secret      =  bin2hex(openssl_random_pseudo_bytes(4));
```
Comment (ramth05) [inline woo-razorpay.php]: @jampulanaveen Can we use bugger length secrets here also if it must have capital, small case letters and numbers.

**Review by ramth05 — CHANGES_REQUESTED** [review]
@jampulanaveen , don't see any list for the supported events.

File: woo-razorpay.php [hunk 6]
```diff
@@ -57,6 +57,13 @@ class WC_Razorpay extends WC_Payment_Gateway
         const DEFAULT_DESCRIPTION            = 'Pay securely by Credit or Debit card or Internet Banking through Razorpay.';
         const DEFAULT_SUCCESS_MESSAGE        = 'Thank you for shopping with us. Your account has been charged and your transaction is successful. We will be processing your order soon.';
 
+        protected $supportedWebhookEvents         = array(
+
+            'payment.authorized' => true,
+            'refund.created' => true
+
+        );
+
```
Comment (ramth05) [inline woo-razorpay.php:66]: @jampulanaveen , remove items from $visibleSettings list.

**Comment by ramth05** [comment]
@jampulanaveen , please verify if how we are going to handle subscription based events in this.

File: woo-razorpay.php [hunk 7]
```diff
@@ -361,64 +342,53 @@ public function autoEnableWebhook()
...
+            ];
+            
             if(count($webhook) > 0)
             {
                 foreach ($webhook['items'] as $key => $value)
                 {
                     if($value['url'] === $webhookUrl)
                     {
+                        $newEvents = [];
+                        foreach($value['events'] as $evntkey => $evntval)
```
Comment (ramth05) [inline woo-razorpay.php:368]: space before `(`

File: woo-razorpay.php [hunk 8]
```diff
@@ -610,6 +580,15 @@ private function getRedirectUrl($orderId)
          */
         protected function getRazorpayPaymentParams($orderId)
         {
+            $getWebhookFlag =  get_option('webhook_enable_flag');
+
+           if(!empty($getWebhookFlag)){
```
Comment (ramth05) [inline woo-razorpay.php:585]: start it with new line `{`

File: woo-razorpay.php [hunk 9]
```diff
@@ -610,6 +580,15 @@ private function getRedirectUrl($orderId)
...
+               
+                if($getWebhookFlag + 86400 < time()){
```
[the remark this answers is not in the document]
Comment (jampulanaveen) [reply woo-razorpay.php]: done

File: woo-razorpay.php (hunk 8 above)
Comment (jampulanaveen) [reply woo-razorpay.php]: done, ok

File: woo-razorpay.php (hunk 7 above)
Comment (jampulanaveen) [reply woo-razorpay.php]: done

File: woo-razorpay.php [hunk 10]
```diff
@@ -57,6 +57,16 @@ class WC_Razorpay extends WC_Payment_Gateway
         const DEFAULT_DESCRIPTION            = 'Pay securely by Credit or Debit card or Internet Banking through Razorpay.';
         const DEFAULT_SUCCESS_MESSAGE        = 'Thank you for shopping with us. Your account has been charged and your transaction is successful. We will be processing your order soon.';
 
+        protected $supportedWebhookEvents = array(
+            'payment.authorized',
```
Comment (ramth05) [inline woo-razorpay.php:61]: @jampulanaveen , missing current supported events?

**Review by nikhilsharma1100 — APPROVED** [review]
LGTM

File: woo-razorpay.php (hunk 10 above)
Comment (jampulanaveen) [reply woo-razorpay.php:61]: updated

**Pull request merged**